File: python/openharness/tools.py
"""Custom tools — Python functions the agent can call during a session.

Use :func:`tool` to decorate a Python callable, then pass the decorated
function (or any regular callable) to :func:`query` or
:class:`OpenHarnessClient`. The SDK spins up an in-process MCP server,
registers your functions as MCP tools, and wires the spawned `oh` CLI
to call back into it.

Example::

    import asyncio
    from openharness import OpenHarnessClient, tool

    @tool
    async def get_weather(city: str) -> str:
        '''Fetch the current weather for a city.'''
        return f"Sunny in {city}, 22°C"

    async def main() -> None:
        async with OpenHarnessClient(tools=[get_weather]) as client:
            async for event in await client.send("What's the weather in Paris?"):
                print(event)

    asyncio.run(main())
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any, TypeVar, overload

__all__ = ["is_tool", "tool", "tool_description", "tool_name"]

F = TypeVar("F", bound=Callable[..., Any])

# Attribute keys used to stamp decorated functions.
_IS_TOOL_ATTR = "_oh_is_tool"
_TOOL_NAME_ATTR = "_oh_tool_name"
_TOOL_DESCRIPTION_ATTR = "_oh_tool_description"


@overload
def tool(fn: F, /) -> F: ...


@overload
def tool(
    *,
    name: str | None = None,
    description: str | None = None,
) -> Callable[[F], F]: ...


def tool(
    fn: F | None = None,
    /,
    *,
    name: str | None = None,
    description: str | None = None,
) -> F | Callable[[F], F]:
    """Mark a function as an OH-callable tool.

    Can be used with or without arguments::

        @tool
        def a(x: int) -> int: ...

        @tool(name="custom-name", description="override")
        def b(x: int) -> int: ...

    Without the decorator, a plain callable still works — decoration is
    purely optional metadata for overriding the inferred name and
    description.

    :param name: Override the tool name exposed to the agent.
        Defaults to ``fn.__name__``.
    :param description: Override the tool description. Defaults to
        the first line of ``fn.__doc__``.
    """

    def _wrap(f: F) -> F:
        setattr(f, _IS_TOOL_ATTR, True)
        if name is not None:
            setattr(f, _TOOL_NAME_ATTR, name)
        if description is not None:
            setattr(f, _TOOL_DESCRIPTION_ATTR, description)
        return f

    if fn is None:
        return _wrap
    return _wrap(fn)


def is_tool(fn: Callable[..., Any]) -> bool:
    """Return True if ``fn`` was decorated with :func:`tool`."""
    return getattr(fn, _IS_TOOL_ATTR, False) is True


def tool_name(fn: Callable[..., Any]) -> str:
    """Return the tool name for ``fn`` (custom or inferred from ``__name__``)."""
    override = getattr(fn, _TOOL_NAME_ATTR, None)
    if isinstance(override, str):
        return override
    return getattr(fn, "__name__", "tool")


def tool_description(fn: Callable[..., Any]) -> str:
    """Return the tool description (custom or first docstring line)."""
    override = getattr(fn, _TOOL_DESCRIPTION_ATTR, None)
    if isinstance(override, str):
        return override
    doc = getattr(fn, "__doc__", None)
    if isinstance(doc, str) and doc.strip():
        return doc.strip().splitlines()[0]
    return ""

File: packages/sdk/src/tools.ts
/**
 * Custom tools — TypeScript functions the agent can call during a session.
 *
 * Pass an array of `ToolDefinition`s to `query()` or `OpenHarnessClient` via
 * `tools: [...]`. The SDK starts an in-process MCP HTTP server, registers
 * each tool, and writes an ephemeral `.oh/config.yaml` pointing the spawned
 * `oh` CLI at it.
 *
 * Mirrors `python/openharness/tools.py`. Python uses an optional `@tool`
 * decorator over plain callables; TS uses an explicit `ToolDefinition`
 * object with a Zod input schema.
 *
 * @example
 * ```ts
 * import { z } from "zod";
 * import { tool } from "@zhijiewang/openharness-sdk";
 *
 * const getWeather = tool({
 *   name: "get_weather",
 *   description: "Fetch the current weather for a city.",
 *   inputSchema: z.object({ city: z.string() }),
 *   handler: async ({ city }) => `Sunny in ${city}, 22°C`,
 * });
 * ```
 */

import type { z } from "zod";

// The default type parameter uses `z.ZodObject<any>` (rather than the tighter
// `z.ZodObject<z.ZodRawShape>`) so that `ToolDefinition[]` and `OpenHarnessOptions.tools`
// can hold a collection of tools with different concrete input shapes. Without
// this, the handler's parameter contravariance prevents assigning, e.g.,
// `ToolDefinition<ZodObject<{ msg: ZodString }>>` to the array's element type.
// Individual tool authors still get strong typing via the `tool()` helper.
export interface ToolDefinition<S extends z.ZodObject<z.ZodRawShape> = z.ZodObject<any>> {
  /** Tool name as exposed to the agent. Should be a stable identifier. */
  name: string;
  /** One-line human description. Shown to the model when it considers calling the tool. */
  description?: string;
  /** Zod object schema describing the tool's input. Validated by the MCP runtime. */
  inputSchema: S;
  /**
   * Tool implementation. Called with the validated input. May return any value;
   * non-string return values are JSON-stringified before being sent back to the
   * agent.
   */
  handler: (input: z.infer<S>) => unknown | Promise<unknown>;
}

/**
 * Identity helper that asserts the shape of a `ToolDefinition` and provides
 * useful inference. Equivalent to a plain object literal but catches typos
 * eagerly and surfaces nicer errors when a required field is missing.
 */
export function tool<S extends z.ZodObject<z.ZodRawShape>>(def: ToolDefinition<S>): ToolDefinition<S> {
  if (!def || typeof def !== "object") throw new TypeError("tool() requires a definition object");
  if (!def.name || typeof def.name !== "string") throw new TypeError("tool() requires a 'name' string");
  if (!def.inputSchema) throw new TypeError(`tool('${def.name}') requires an 'inputSchema' (Zod object schema)`);
  if (typeof def.handler !== "function") throw new TypeError(`tool('${def.name}') requires a 'handler' function`);
  return def;
}
